### PR TITLE
Fix resolver by id and slug for product and product variant

### PR DESCRIPTION
**Pull request by d-wysocki**
I want to merge this change because...I want to fix resolvers by id and slug for product and product variant.

# Impact

* [ ] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migration files are up to date
* [ ] The changes are tested
* [ ] GraphQL schema and type definitions are up to date
* [x] Changes are mentioned in the changelog

### CODE REVIEW

File: saleor/graphql/product/tests/test_product.py [hunk 1]
```diff
@@ -6373,6 +6373,144 @@ def test_product_update_variants_names(mock__update_variants_names, product_type
...
+    product_channel_listing.visible_in_listings = False
+    product_channel_listing.save()
+
+    variables = {"slug": product.slug, "channel": channel_USD.slug}
+    response = user_api_client.post_graphql(query, variables)
+    content = get_graphql_content(response)
+    assert content["data"]["product"] is None
+
+
+def test_unavailable_product_variant_by_id_as_staff(
```
Comment (korycins) [inline saleor/graphql/product/tests/test_product.py:6465]: IMO, for all your tests which check the queries for staff, I would add a pytest.parametrize to cover the app case also.
Something like this should work:
```
@pytest.mark.parametrize("requestor", ["staff", "app"])
def test_exclude_products_for_shipping_method_only_products(
    requestor,
    app_api_client,
    shipping_method,
    product_list,
    staff_api_client,
    permission_manage_shipping,
):
    api = staff_api_client if requestor == "staff" else app_api_client
    ```

Comment (patrys) [reply saleor/graphql/product/tests/test_product.py:6465]: Please, no `if`s in tests. Those are two separate test cases.

**Review by IKarbowiak — APPROVED** [review]

File: saleor/graphql/product/resolvers.py [hunk 2]
```diff
@@ -47,20 +47,28 @@ def resolve_digital_contents(_info):
...
-        .first()
-    )
+def resolve_product_by_id(
+    info, id, channel_slug, requestor, requestor_has_access_to_all
+):
+    visible_products = models.Product.objects.visible_to_user(requestor, channel_slug)
+    if not requestor_has_access_to_all:
+        visible_products = visible_products.annotate_visible_in_listings(
+            channel_slug
+        ).exclude(visible_in_listings=False)
```
Comment (fowczarek) [inline saleor/graphql/product/resolvers.py:57]: Maybe we should make a separate function from these lines?

We have the same block of code in many resolvers.

**Review by fowczarek — CHANGES_REQUESTED** [review]

**Review by fowczarek — APPROVED** [review]

File: saleor/graphql/product/resolvers.py (hunk 2 above)
Comment (d-wysocki) [reply saleor/graphql/product/resolvers.py:57]: I will do that

File: saleor/graphql/product/tests/test_product.py (hunk 1 above)
Comment (korycins) [reply saleor/graphql/product/tests/test_product.py:6465]: So it should be tested for customer, staff_user and app.

Comment (d-wysocki) [reply saleor/graphql/product/tests/test_product.py:6511]: I added tests for app.

File: saleor/graphql/product/mutations/products.py [hunk 3]
```diff
@@ -989,7 +989,6 @@ def perform_mutation(cls, _root, info, **data):
         )
 
         response = super().perform_mutation(_root, info, **data)
-
```
Comment (fowczarek) [inline saleor/graphql/product/mutations/products.py:992]: I think we shouldn't refactor files not related to this PR.

**Review by fowczarek — APPROVED** [review]

**Review by maarcingebala — APPROVED** [review]

**Comment by maarcingebala** [comment]
@d-wysocki A note for the future PRs - please provide more accurate descriptions in the PR and in the changelog. It says that the PR fixes the resolver, but it doesn't say anything about what was broken and what exactly is fixed here.

**Pull request merged**